exporter/awsemfexporter/metric_translator.go

```diff
@@ -118,6 +134,278 @@ func (dps DoubleHistogramDataPointSlice) At(i int) DataPoint {
+		} else if svcNsOk && serviceNamespace.Type() == pdata.AttributeValueSTRING {
+			namespace = serviceNamespace.StringVal()
+		}
+	}
+
+	if len(namespace) == 0 {
+		namespace = defaultNameSpace
+	}
+	return namespace
+}
+
+// TranslateOtToGroupedMetric converts OT metrics to GroupedMetric format
+func TranslateOtToGroupedMetric(metric pdata.Metrics, config *Config) (map[string]*GroupedMetric, int) {
+	totalDroppedMetrics := 0
+	groupedMetricMap := make(map[string]*GroupedMetric)
+	var instrumentationLibName string
+	rms := metric.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rm := rms.At(i)
+		if rm.IsNil() {
+			continue
+		}
+		namespace := getNamespace(&rm, config.Namespace)
+		ilms := rm.InstrumentationLibraryMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ilm := ilms.At(j)
+			if ilm.IsNil() {
+				continue
+			}
+
+			if ilm.InstrumentationLibrary().IsNil() {
+				instrumentationLibName = noInstrumentationLibraryName
+			} else {
+				instrumentationLibName = ilm.InstrumentationLibrary().Name()
+			}
+
+			metrics := ilm.Metrics()
+			if metrics.Len() > 0 {
+				for k := 0; k < metrics.Len(); k++ {
+					metric := metrics.At(k)
+					if metric.IsNil() {
+						totalDroppedMetrics++
+						continue
+					}					
+					getGroupedMetrics(&metric, namespace, instrumentationLibName, groupedMetricMap, config)
+				}
+			}
+		}
+	}
+	return groupedMetricMap, totalDroppedMetrics
+}
+
+
+func getGroupedMetrics(metric *pdata.Metric, namespace string, instrumentationLibName string, groupedMetricMap map[string]*GroupedMetric, config *Config) (groupedMetrics []*GroupedMetric){
+	if metric == nil {
+		return
+	}
+
+	// Retrieve Datapoints
+	var dps DataPoints
+	switch metric.DataType() {
+		case pdata.MetricDataTypeIntGauge:
+			dps = IntDataPointSlice{metric.IntGauge().DataPoints()}
+		case pdata.MetricDataTypeDoubleGauge:
+			dps = DoubleDataPointSlice{metric.DoubleGauge().DataPoints()}
+		case pdata.MetricDataTypeIntSum:
+			dps = IntDataPointSlice{metric.IntSum().DataPoints()}
+		case pdata.MetricDataTypeDoubleSum:
+			dps = DoubleDataPointSlice{metric.DoubleSum().DataPoints()}
+		case pdata.MetricDataTypeDoubleHistogram:
+			dps = DoubleHistogramDataPointSlice{metric.DoubleHistogram().DataPoints()}
+		default:
+			config.logger.Warn("Unhandled metric data type.",
+				zap.String("DataType", metric.DataType().String()),
+				zap.String("Name", metric.Name()),
+				zap.String("Unit", metric.Unit()),)
+			return
+	}
+
+	if dps.Len() == 0 {
+		return
+	}
+
+	for m := 0; m < dps.Len(); m++ {
+		dp := dps.At(m)
+		if dp.IsNil() {
+			continue
+		}
+
+		fields := make(map[string]interface{})
+		s := []string{nameSpace, namespace, OTellibDimensionKey, instrumentationLibName}
+		fields[nameSpace] = namespace
+		fields[OTellibDimensionKey] = instrumentationLibName
+
+		dp.LabelsMap().ForEach(func(k string, v string) {
+			fields[k] = v
+			if (v != noInstrumentationLibraryName) {
+				s =append(s, k, v)
+			}
+		})
+
+		sort.Strings(s)
```

Comment:
I think this logic will generate ambiguous keys. You'll need to pair up key-value pairs and then sort the pairs. Also, since you're using a slice to generate the key, this might have unintentional duplicate keys since you're not directly converting from a map.